File: Python/repeated_substring_pattern.py
# https://leetcode.com/problems/repeated-substring-pattern/description/
from xdg.Mime import app_exe

from utils.examination import Examiner, Testcase, Logger


class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        l: int = len(s)
        if 0 == 1 & l and s[:(l // 2)] == s[(l // 2):]:
            return True

        i: int = 0
        if 1 == 1 & l:
            i = 1
        left: int = (l - i) // 2
        right: int = left + i
        while 1 <= left:
            # Logger.log(i, left, right, s[:left], s[right:], s[left:right], s[:i], s[-i:])
            if s[:left] == s[right:] and s[left:right] == s[:i] == s[-i:]:
                return True
            i += 2
            left: int = (l - i) // 2
            right: int = left + i
        return False


def examine() -> None:
    Examiner(Solution(), 'repeatedSubstringPattern').run((
        Testcase(('babba',), False),
        Testcase(('ababab',), True),
        Testcase(('abaaaabaaaabaaa',), True),
        Testcase(('aaa',), True),
        Testcase(('abab',), True),
        Testcase(('aba',), False),
        Testcase(('abcabcabcabc',), True),
        Testcase(('a',), False),
        Testcase(('abaababaab',), True),
    ))

File: TypeScript/solutions/repeated-substring-pattern.ts
// https://leetcode.com/problems/repeated-substring-pattern/description/
import { Examiner, Testcase } from '../utils';


function repeatedSubstringPattern(s: string): boolean {
    const l = s.length;
    if ((l & 1) === 0 && s.slice(0, l / 2) === s.slice(l / 2)) return true;

    let i = l % 2 === 1 ? 1 : 0;
    let left = Math.floor((l - i) / 2);
    let right = left + i;
    while (left >= 1) {
        if (s.slice(0, left) === s.slice(right) && s.slice(left, right) === s.slice(0, i) && s.slice(0, i) === s.slice(-i || l))
            return true;
        i += 2;
        left = Math.floor((l - i) / 2);
        right = left + i;
    }
    return false;
}

export function examine(): void {
    new Examiner(repeatedSubstringPattern).run([
        new Testcase(['babba'], false),
        new Testcase(['ababab'], true),
        new Testcase(['abaaaabaaaabaaa'], true),
        new Testcase(['aaa'], true),
        new Testcase(['abab'], true),
        new Testcase(['aba'], false),
        new Testcase(['abcabcabcabc'], true),
        new Testcase(['a'], false),
        new Testcase(['abaababaab'], true),
    ]);
}
